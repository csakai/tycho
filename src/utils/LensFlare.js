import * as THREE from 'three';
import Constants from '../constants';

export default class LensFlare extends THREE.LensFlare {

    constructor(camera) {
        super();
        this.render();
        this.position.set(0, 0, 0);
        this.camera = camera;
    }

    /**
     * Texture loader.
     *
     * @type {THREE.TextureLoader}
     */
    textureLoader = new THREE.TextureLoader()

    /**
     * Lens flare color.
     *
     * @type {THREE.Color}
     */
    color = new THREE.Color(Constants.WebGL.Sunlight.COLOR)

    /**
     * Blending mode.
     *
     * @type {Number}
     */
    blending = THREE.AdditiveBlending

    /**
     * Adds a lens flare texture to the set.
     *
     * @param {Object} options
     * @param {String} options.url - file name of texture
     * @param {Number} options.diameter - size, in pixels, of texture
     * @param {Number} options.distance - relative distance from light source to camera [0,1]
     */
    addLensFlare = ({url, diameter, distance}) => {
        this.textureLoader.load(`/static/textures/lensflares/${url}`, (texture) => {
            this.add(texture, diameter, distance, this.blending, this.color);
        });
    }

    /**
     * Renders the lens flare.
     */
    render = () => {
        Constants.WebGL.LENS_FLARES.forEach(this.addLensFlare);
    }

    /**
     * Updates lens flares update positions on all flares based on the screen position.
     * Set myLensFlare.customUpdateCallback to alter the flares in your project specific way.
     *
     * @overrides THREE.LensFlare.updateLensFlares
     */
    updateLensFlares = () => {
        let flare;

        const fl = this.lensFlares.length;
        const vecX = -this.positionScreen.x * 2;
        const vecY = -this.positionScreen.y * 2;

        const cameraDistance = this.camera.position.length();
        const percentDistance = 1 - cameraDistance / Constants.WebGL.Camera.MAX_DISTANCE;
        const isZoomedTooFar = percentDistance < Constants.WebGL.LENS_FLARE_MAX_DISTANCE;

        for(let f = 0; f < fl; f++) {
            flare = this.lensFlares[f];

            flare.x = this.positionScreen.x + vecX * flare.distance;
            flare.y = this.positionScreen.y + vecY * flare.distance;

            flare.wantedRotation = flare.x * Math.PI * 0.25;
            flare.rotation += (flare.wantedRotation - flare.rotation) * 0.25;

            if (flare.distance === 0.0) {
                flare.scale = percentDistance;
            } else {
                flare.scale = isZoomedTooFar ? 0 : 1;
            }
        }
    }
}
