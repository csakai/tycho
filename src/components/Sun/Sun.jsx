import React from 'react';
import PropTypes from 'prop-types';
import Constants from '../../constants';
import LensFlare from '../../utils/LensFlare';

export default class Sun extends React.Component {

    static propTypes = {
        camera: PropTypes.object
    }

    componentDidMount = () => {
        this.refs.sun.add(new LensFlare(this.props.camera));
    }

    render() {
        return (
            <group ref="sun">
                <pointLight
                    color={Constants.WebGL.Sunlight.COLOR}
                    intensity={Constants.WebGL.Sunlight.INTENSITY}
                    distance={Constants.WebGL.Sunlight.DISTANCE}
                />
            </group>
        );
    }
}
