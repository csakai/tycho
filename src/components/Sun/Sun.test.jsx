import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {PerspectiveCamera} from 'three';
import Sun from './Sun';
import LensFlare from '../../utils/LensFlare';

describe.only('The Sun', () => {
    const camera = new PerspectiveCamera();
    let component, sun;

    beforeEach(() => {
        component = shallow((
            <Sun camera={camera} />
        ));
        sun = component.instance();
    });

    describe('componentDidMount()', () => {
        const children = [];

        beforeEach(() => {
            sun.props = {camera};
            sun.refs = {
                sun: {
                    add: (e) => children.push(e)
                }
            };
        });

        it('should add a new instance of LensFlare', () => {
            sun.componentDidMount();

            expect(children).toHaveLength(1);
            expect(children[0]).toBeInstanceOf(LensFlare);
        });

        it('should add a new element to the Sun ref via add()', () => {
            const spy = jest.spyOn(sun.refs.sun, 'add');

            sun.componentDidMount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    describe('render()', () => {
        it('should render the Sun', () => {
            expect(toJson(component)).toMatchSnapshot();
        });
    });
});
