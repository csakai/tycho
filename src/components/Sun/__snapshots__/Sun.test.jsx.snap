// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Sun render() should render the Sun 1`] = `
<group>
  <pointLight
    color={16777215}
    distance={20000}
    intensity={0.95}
  />
</group>
`;
